pkg/external_repos/introspect.go

```diff
@@ -0,0 +1,123 @@
+package external_repos
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/content-services/content-sources-backend/pkg/config"
+	"github.com/content-services/content-sources-backend/pkg/dao"
+	"github.com/content-services/content-sources-backend/pkg/db"
+	"github.com/content-services/content-sources-backend/pkg/models"
+	"github.com/content-services/yummy/pkg/yum"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	RhCdnHost = "cdn.redhat.com"
+)
+
+// IntrospectUrl Fetch the metadata of a url and insert RPM data
+//  Returns the number of new RPMs inserted system-wide and any error encountered
+func IntrospectUrl(url string) (int64, error) {
+	err, publicRepo := dao.GetPublicRepositoryDao(db.DB).FetchForUrl(url)
+	rpmDao := dao.GetRpmDao(db.DB, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	return Introspect(publicRepo, rpmDao)
+}
+
+// IsRedHat returns if the url is a 'cdn.redhat.com' url
+func IsRedHat(url string) bool {
+	return strings.Contains(url, RhCdnHost)
+}
+
+// Introspect introspects a dao.PublicRepository with the given RpmDao
+// 	inserting any needed RPMs and adding and removing associations to the repository
+//  Returns the number of new RPMs inserted system-wide and any error encountered
+func Introspect(repo dao.PublicRepository, rpm dao.RpmDao) (int64, error) {
+	var (
+		client http.Client
+		err    error
+		pkgs   []yum.Package
+	)
+	log.Debug().Msg("Introspecting " + repo.URL)
+
+	if client, err = httpClient(IsRedHat(repo.URL)); err != nil {
+		return 0, err
+	}
+	if pkgs, err = yum.ExtractPackageData(client, repo.URL); err != nil {
+		return 0, err
+	}
+	return rpm.InsertForRepository(repo.UUID, pkgs)
+}
+
+// IntrospectAll introspects all repositories
+//  Returns the number of new RPMs inserted system-wide and all errors encountered
+func IntrospectAll() (int64, []error) {
+	var repos []models.Repository
+	var errors []error
+	var total int64
+	var count int64
+	var err error
+	thisdb := db.DB
+	rpmDao := dao.GetRpmDao(thisdb, nil)
+	result := thisdb.Find(&repos)
```

Comment:
Can we move this Find() to the repository dao layer?

Reply:
lets do this as a another pr, and rename some of the dao layer.  I'll open an issue